SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+            }
+            OWSMessageProcessingJob *job1 = (OWSMessageProcessingJob *)object1;
+
+            if (![object2 isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object2, collection2) return NSOrderedSame;
+            }
+            OWSMessageProcessingJob *job2 = (OWSMessageProcessingJob *)object2;
+
+            return [job1.createdAt compare:job2.createdAt];
+        }];
+
+    YapDatabaseViewGrouping *grouping =
+        [YapDatabaseViewGrouping withObjectBlock:^NSString *_Nullable(YapDatabaseReadTransaction *_Nonnull transaction,
+            NSString *_Nonnull collection,
+            NSString *_Nonnull key,
+            id _Nonnull object) {
+            if (![object isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object, collection) return nil;
+            }
+
+            // Arbitrary string - all in the same group. We're only using the view for sorting.
+            return OWSMessageProcessingJobFinderExtensionGroup;
+        }];
+
+    YapDatabaseViewOptions *options = [YapDatabaseViewOptions new];
+    options.allowedCollections =
+        [[YapWhitelistBlacklist alloc] initWithWhitelist:[NSSet setWithObject:[OWSMessageProcessingJob collection]]];
+
+    return [[YapDatabaseView alloc] initWithGrouping:grouping sorting:sorting versionTag:@"1" options:options];
+}
+
+
++ (void)syncRegisterDatabaseExtension:(YapDatabase *)database
+{
+    YapDatabaseView *existingView = [database registeredExtension:OWSMessageProcessingJobFinderExtensionName];
+    if (existingView) {
+        // already initialized
+        return;
+    }
+    [database registerExtension:[self databaseExension] withName:OWSMessageProcessingJobFinderExtensionName];
+}
+
+@end
+
+#pragma mark - Queue Processing
+
+@interface OWSMessageProcessingQueue : NSObject
+
+@property (nonatomic, readonly) TSMessagesManager *messagesManager;
+@property (nonatomic, readonly) OWSMessageProcessingJobFinder *finder;
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+@implementation OWSMessageProcessingQueue
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder
+{
+    OWSSingletonAssert();
+
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _messagesManager = messagesManager;
+    _finder = finder;
+
+    return self;
+}
+
+#pragma mark - instance methods
+
+- (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    [self.finder addJobForEnvelope:envelope];
+}
+
+- (void)drainQueue
```

Comment:
Yeah, it would make sense as a method name if its was purely recursive, but since it dispatches async, I tend to think the name should reflect the effect or work done by a single invocation.  Having said that, I was just asking.  Fine to leave as it is.